hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,198 @@
 
 import asyncio
 import concurrent.futures._base as base
+import ctypes
 import multiprocessing as mp
 import multiprocessing.connection
-import time
-from functools import lru_cache
-from typing import Optional, Tuple, Generic, TypeVar
+import os
+import threading
+from typing import Tuple, Generic, TypeVar, Dict, Optional, Any, Callable
+from hivemind.utils.logging import get_logger
 
-from hivemind.utils.threading import run_in_background
 
-ResultType = TypeVar('ResultType')
+logger = get_logger(__name__)
 
+# flavour types
+ResultType = TypeVar('ResultType')
+PID, UID, State, PipeEnd = int, int, Any, mp.connection.Connection
+ALL_STATES = base.PENDING, base.RUNNING, base.FINISHED, base.CANCELLED, base.CANCELLED
```

Comment:
Typo: `base.CANCELLED` twice instead of `base.CANCELLED, base.CANCELLED_AND_NOTIFIED`